hypothesis-python/src/hypothesis/internal/conjecture/datatree.py

```diff
@@ -305,6 +764,97 @@ def simulate_test_function(self, data):
     def new_observer(self):
         return TreeRecordingObserver(self)
 
+    def _draw(self, ir_type, kwargs, *, random, forced=None):
+        # we should possibly pull out BUFFER_SIZE to a common file to avoid this
+        # circular import.
+        from hypothesis.internal.conjecture.engine import BUFFER_SIZE
+
+        cd = ConjectureData(max_length=BUFFER_SIZE, prefix=b"", random=random)
+        draw_func = getattr(cd, f"draw_{ir_type}")
+
+        value = draw_func(**kwargs, forced=forced)
+        buf = cd.buffer
+
+        # using floats as keys into branch.children breaks things, because
+        # e.g. hash(0.0) == hash(-0.0) would collide as keys when they are
+        # in fact distinct child branches.
+        # To distinguish floats here we'll use their bits representation. This
+        # entails some bookkeeping such that we're careful about when the
+        # float key is in its bits form (as a key into branch.children) and
+        # when it is in its float form (as a value we want to write to the
+        # buffer), and converting between the two forms as appropriate.
+        if ir_type == "float":
+            value = float_to_int(value)
+        return (value, buf)
+
+    def _get_children_cache(self, ir_type, kwargs, *, key):
+        # cache the state of the children generator per node/branch (passed as
+        # `key` here), such that we track which children we've already tried
+        # for this branch across draws.
+        # We take advantage of python generators here as one-way iterables,
+        # so each time we iterate we implicitly store our position in the
+        # children generator and don't re-draw children. `children` is the
+        # concrete list of children draw from the generator that we will work
+        # with. Whenever we need to top up this list, we will draw a new value
+        # from the generator.
+        if key not in self._children_cache:
+            generator = all_children(ir_type, kwargs)
+            children = []
+            rejected = set()
+            self._children_cache[key] = (generator, children, rejected)
+
+        return self._children_cache[key]
+
+    def _draw_from_cache(self, ir_type, kwargs, *, key, random):
+        (generator, children, rejected) = self._get_children_cache(
+            ir_type, kwargs, key=key
+        )
+        # Keep a stock of 100 potentially-valid children at all times.
+        # This number is chosen to balance memory/speed vs randomness. Ideally
+        # we would sample uniformly from all not-yet-rejected children, but
+        # computing and storing said children is not free.
+        # no-branch because coverage of the fall-through case here is a bit
+        # annoying.
+        if len(children) < 100:  # pragma: no branch
+            for v in generator:
+                if ir_type == "float":
+                    v = float_to_int(v)
```

Comment:
I had intended for this to be covered by `test_can_generate_hard_floats`, but forgot it can be flaky due to lucking into the correct value (which avoids our child cache logic).